File: collector/lib/codeg.js
// Webhooks discover connections; read-only event streams reconcile confirmations. SQLite is used
// only for API credentials and one keyed metadata lookup after an event.
import path from 'node:path';
import fs from 'node:fs/promises';
import { randomUUID } from 'node:crypto';
import { question, questionDetails } from './codex.js';
let DatabaseSync;
try { ({ DatabaseSync } = await import('node:sqlite')); } catch {}
export const SOURCE = 'codeg';
export const CODEG_EVENTS = ['user_prompt_sent','question_request','permission_request','turn_complete','error'];
export function defaultCodegDbPaths(home) {
  return [path.join(home,'Library/Application Support/app.codeg/codeg.db'),path.join(home,'Library/Application Support/codeg/codeg.db'),path.join(home,'.local/share/codeg/codeg.db')];
}
export function mergeCodegWebhooks(existing, owned, url) {
  if (!Array.isArray(existing) || existing.some(w=>!w || typeof w.url!=='string' || typeof w.enabled!=='boolean')) throw Error('Codeg Webhook 配置无效，未覆盖');
  const next=existing.filter(w=>!owned.includes(w.url) && w.url!==url);
  if(url)next.push({url,enabled:true});
  return next;
}
export function askFromSnapshot(snap) {
  if(snap?.pending_question)return snap.pending_question;
  if(snap?.pending_plan_approval)return {approval_id:snap.pending_plan_approval.approval_id,questions:[{question:snap.pending_plan_approval.plan_markdown,options:[{label:'批准'},{label:'拒绝'}]}]};
  if(snap?.pending_permission){const p=snap.pending_permission;return {request_id:p.request_id,questions:[{question:p.tool_call?.title||p.tool_call?.name||'需要你的许可',options:(p.options||[]).map(o=>({label:o.name||o.label||'',description:o.kind||''}))}]};}
  return null;
}
export class CodegHooks {
  constructor(hub,{home,dbPaths=defaultCodegDbPaths(home),fetchImpl=globalThis.fetch,now=Date.now,WebSocketImpl=globalThis.WebSocket}={}) {
    Object.assign(this,{hub,home,dbPaths,fetchImpl,now,WebSocketImpl});
    this.url='';this.auth=null;this.registered=false;this.nextAttempt=0;this.enabled=true;this.sequence=0;this.connections=new Map();this.ignoredConnections=new Set();
    this.streams=new Map();
    this.stateFile=path.join(home,'.agent-studio/codeg-webhook-node.json');
  }
  open() {
    if(!DatabaseSync)throw Error('Codeg 配置需要 Node 22.13+');
    for(const p of this.dbPaths){try{return new DatabaseSync(p,{readOnly:true});}catch{}}
    throw Error('未找到 Codeg 数据库，无法读取 Web Service 配置');
  }
  credentials() {
    const db=this.open();try {
      const meta=Object.fromEntries(db.prepare("SELECT key,value FROM app_metadata WHERE key IN ('web_service_port','web_service_token')").all().map(r=>[r.key,r.value]));
      const port=Number(meta.web_service_port||3080),token=meta.web_service_token;
      if(!Number.isInteger(port)||port<1||port>65535||!token)throw Error('请启用 Codeg Web Service');
      return {port,token};
    } finally {db.close();}
  }
  async post(name,body={}) {
    const response=await this.fetchImpl(`http://127.0.0.1:${this.auth.port}/api/${name}`,{method:'POST',headers:{'Content-Type':'application/json',Authorization:`Bearer ${this.auth.token}`},body:JSON.stringify(body),signal:AbortSignal.timeout(1500)});
    if(!response.ok)throw Error(`Codeg API HTTP ${response.status}`);
    return response.json();
  }
  async install(base) {this.url=`${base}/api/codeg-webhook/${randomUUID()}`;this.registered=false;this.nextAttempt=0;await this.poll();}
  async poll() {
    // Only failed configuration is retried. No timer reads session state.
    if(!this.url){this.hub.health(SOURCE,'partial','Webhook 接收入口尚未启动');return;}
    if(this.registered || this.now()<this.nextAttempt)return;
    this.nextAttempt=this.now()+60000;
    try {
      let owned=[];try{owned=JSON.parse(await fs.readFile(this.stateFile,'utf8')).owned;if(!Array.isArray(owned))throw Error('invalid');}catch(e){if(e.code!=='ENOENT')throw Error('Codeg 注册记录不可读，未覆盖');}
      if(!this.enabled&&!owned.length){this.registered=true;return;}
      this.auth=this.credentials();
      const existing=await this.post('get_chat_event_webhooks');
      const next=mergeCodegWebhooks(existing,owned,this.enabled?this.url:'');
      if(this.enabled){
        const filter=await this.post('get_chat_event_filter');
        if(filter!==null && (!Array.isArray(filter)||filter.some(e=>typeof e!=='string')))throw Error('Codeg 事件配置无效，未覆盖');
        const current=filter??CODEG_EVENTS.filter(e=>e!=='user_prompt_sent');
        if(CODEG_EVENTS.some(e=>!current.includes(e))){
          // The filter is global: don't turn on new messages to other sinks.
          const channels=await this.post('list_chat_channels');
          if(!Array.isArray(channels)||channels.some(c=>c.enabled)||next.some(w=>w.url!==this.url&&w.enabled))throw Error('Codeg 全局事件开关影响其他推送目标；请先在 Codeg 启用所需的五类事件');
          await this.post('set_chat_event_filter',{filter:[...new Set([...current,...CODEG_EVENTS])]});
        }
      }
      // Journal both URLs before mutation; retry uncertain writes without duplicates.
      const url=this.enabled?this.url:'';
      if(url&&!owned.includes(url))owned.push(url);
      await fs.mkdir(path.dirname(this.stateFile),{recursive:true});
      const save=async value=>{const tmp=this.stateFile+'.tmp';await fs.writeFile(tmp,JSON.stringify(value),{mode:0o600});await fs.rename(tmp,this.stateFile);};
      await save({owned});
      if(JSON.stringify(existing)!==JSON.stringify(next))await this.post('set_chat_event_webhooks',{webhooks:next});
      await save({owned:url?[url]:[]});
      this.registered=true;
      this.hub.health(SOURCE,this.enabled?'ok':'disabled',this.enabled?'Webhook 已注册，等待 Codeg 事件（不扫描会话）':'已关闭监听');
    } catch(e) {this.hub.health(SOURCE,this.enabled?'error':'disabled',this.enabled?`Webhook 注册失败：${e.message?.startsWith('Codeg')||e.message?.startsWith('请')||e.message?.startsWith('未')?e.message:'请确认 Codeg Web Service 可用'}`:'已关闭监听；Webhook 注销待重试');}
  }
  async disable(){this.enabled=false;for(const stream of this.streams.values())stream.close();this.streams.clear();this.registered=false;this.nextAttempt=0;await this.poll();}
  metadata(sid) {
    const db=this.open();try{
      const tables=db.prepare("SELECT name FROM sqlite_master WHERE type='table'").all().map(r=>r.name);
      const table=tables.includes('conversation')?'conversation':'conversations';
      const row=db.prepare(`SELECT * FROM ${table} WHERE id=?`).get(sid)||{};
      let cwd=row.origin_cwd||row.cwd||row.workspace||'';
      if(!cwd&&row.folder_id&&tables.includes('folder'))cwd=db.prepare('SELECT path FROM folder WHERE id=?').get(row.folder_id)?.path||'';
      return {title:row.title||'',cwd,agentType:row.agent_type||row.agent||'',externalId:row.external_id||'',folderId:row.folder_id,isSubagent:row.parent_id!=null||row.kind==='delegate'};
    } finally{db.close();}
  }
  async ingestHook(p) {
    if(!this.enabled||p?.source!=='codeg'||!CODEG_EVENTS.includes(p.event)||typeof p.connection_id!=='string'||!p.connection_id.trim()||p.connection_id.length>256)return false;
    if(this.ignoredConnections.has(p.connection_id))return true;
    let snap=null;
    try {this.auth??=this.credentials();snap=await this.post('acp_get_session_snapshot',{connectionId:p.connection_id});}catch{}
    if(!this.enabled)return false;
    const sid=String(snap?.conversation_id??this.connections.get(p.connection_id)??`connection:${p.connection_id}`);
    const provisional=`codeg:connection:${p.connection_id}`;
    if(!sid.startsWith('connection:')){this.connections.set(p.connection_id,sid);this.hub.sessions.delete(provisional);}
    if(this.connections.size>512)this.connections.delete(this.connections.keys().next().value);
    let meta={};if(!sid.startsWith('connection:'))try{meta=this.metadata(sid);}catch{}
    if(meta.isSubagent) {
      // A known child stays ignored even if later snapshot/metadata reads fail.
      this.ignoredConnections.add(p.connection_id);
      this.streams.get(p.connection_id)?.close();this.streams.delete(p.connection_id);
      this.hub.sessions.delete(provisional);
      this.hub.sessions.delete(`codeg:${sid}`);
      return true;
    }
    const existingStream=this.streams.get(p.connection_id);
    if(existingStream?.sid!==sid){existingStream?.close();this.streams.delete(p.connection_id);}
    const active=this.streams.get(p.connection_id);
    // A keyed HTTP response may arrive after newer stream frames. Never let it
    // rewind pending requests or reopen a round already completed by the stream.
    if(active?.seq!=null && Number.isSafeInteger(snap?.event_seq) && snap.event_seq<active.seq)return true;
    if(active?.seq!=null && p.event==='user_prompt_sent' && snap?.status && snap.status!=='prompting' && ['done','error','aborted'].includes(this.hub.sessions.get(`codeg:${sid}`)?.status))return true;
    if(active && active.seq!==null && ['question_request','permission_request'].includes(p.event))return true;
    if(p.event==='turn_complete'&&snap?.status==='prompting')return true;
    const ts=this.now(),seq=++this.sequence;
    const base={source:SOURCE,sessionId:sid,ts,...meta,externalId:snap?.external_id||meta.externalId,folderId:snap?.folder_id??meta.folderId};
    const send=e=>this.hub.ingest({...base,...e});
    const current=this.hub.sessions.get(`codeg:${sid}`);
    if(p.event==='user_prompt_sent'||!current)send({type:'start',roundId:`hook:${ts}:${seq}`,title:meta.title||String(p.body||'Codeg').slice(0,240)});
    if((p.event==='question_request'||p.event==='permission_request') && !['done','error','aborted'].includes(this.hub.sessions.get(`codeg:${sid}`)?.status)) {
      const ask=askFromSnapshot(snap);
      const fields=Array.isArray(p.fields)?p.fields:[];
      // Native webhook fields retain question/option text when snapshot unavailable.
      const text=ask?question(ask):fields.map(f=>String(f.value||'')).filter(Boolean).join('\n')||String(p.body||p.title||'等待确认');
      for(const pending of [...(this.hub.sessions.get(`codeg:${sid}`)?.pending||[])])send({type:'resolve',callId:pending.id});
      send({type:'wait',callId:ask?.question_id||ask?.request_id||ask?.approval_id||`codeg:${p.connection_id}:${seq}`,tool:p.event==='question_request'?'ask':'permission',text,questions:ask?questionDetails(ask):[]});
    } else if(p.event==='turn_complete'||p.event==='error')send({type:'end',status:p.event==='error'?'error':'done'});
    // The keyed snapshot can already be ahead of a delayed webhook.
    if(snap?.status==='prompting')this.reconcileSnapshot(sid,snap);
    this.attachStream(p.connection_id,sid);
    if(Number.isSafeInteger(snap?.event_seq)){const stream=this.streams.get(p.connection_id);if(stream)stream.seq=Math.max(stream.seq??-1,snap.event_seq);}
    this.hub.health(SOURCE,'ok','已收到 Codeg Webhook；确认状态通过实时事件同步');
    return true;
  }
  reconcileSnapshot(sid,snap,authoritative=false) {
    const session=this.hub.sessions.get(`codeg:${sid}`);
    if(!session||['done','error','aborted'].includes(session.status))return;
    const asks=['pending_question','pending_permission','pending_plan_approval'].flatMap(key=>{
      if(!snap[key])return [];
      const ask=askFromSnapshot({[key]:snap[key]});
      const id=ask?.question_id||ask?.request_id||ask?.approval_id;
      return typeof id==='string'&&id?[{id,ask,tool:key==='pending_question'?'ask':'permission'}]:[];
    });
    const send=e=>this.hub.ingest({source:SOURCE,sessionId:sid,ts:this.now(),...e});
    if(authoritative && snap.status==='connected' && ['pending_question','pending_permission','pending_plan_approval'].every(key=>snap[key]==null)){send({type:'end',status:'done'});return;}
    // Only a prompting snapshot proves that an absent request has resumed.
    if(snap.status==='prompting')for(const pending of [...session.pending])if(!asks.some(a=>a.id===pending.id))send({type:'resolve',callId:pending.id});
    for(const {id,ask,tool} of asks)send({type:'wait',callId:id,tool,text:question(ask),questions:questionDetails(ask)});
  }
  applyEnvelope(sid,envelope) {
    const session=this.hub.sessions.get(`codeg:${sid}`);
    if(!session||['done','error','aborted'].includes(session.status))return;
    const send=e=>this.hub.ingest({source:SOURCE,sessionId:sid,ts:this.now(),...e});
    const kind=envelope.type;
    const id=envelope.question_id||envelope.request_id||envelope.approval_id;
    if(['question_resolved','permission_resolved','plan_approval_resolved'].includes(kind)) {
      if(typeof id==='string'&&id)send({type:'resolve',callId:id});
    } else if(['question_request','permission_request','plan_approval_request'].includes(kind)) {
      const key=kind==='question_request'?'pending_question':kind==='permission_request'?'pending_permission':'pending_plan_approval';
      const ask=askFromSnapshot({[key]:envelope});
      if(typeof id==='string'&&id)send({type:'wait',callId:id,tool:kind==='question_request'?'ask':'permission',text:question(ask),questions:questionDetails(ask)});
    } else if(kind==='turn_complete')send({type:'end',status:'done'});
  }
  attachStream(connectionId,sid) {
    if(!this.auth||!this.WebSocketImpl||this.streams.has(connectionId)||!this.enabled)return;
    if(this.streams.size>=512){const key=this.streams.keys().next().value;this.streams.get(key).close();this.streams.delete(key);}
    const subscription=randomUUID();let socket,timer,closed=false,attempt=0;
    const state={sid,seq:null,close:()=>{closed=true;clearTimeout(timer);socket?.close();}};
    this.streams.set(connectionId,state);
    const connect=()=>{
      if(closed||!this.enabled)return;
      const protocols=['codeg-events',`codeg-token.${Buffer.from(this.auth.token.trim()).toString('base64url')}`];
      socket=new this.WebSocketImpl(`ws://127.0.0.1:${this.auth.port}/ws/events`,protocols);
      const current=socket;
      const attach=()=>current.send(JSON.stringify({action:'attach',subscription_id:subscription,connection_id:connectionId,since_seq:state.seq}));
      current.addEventListener('open',attach);
      current.addEventListener('error',()=>{});
      current.addEventListener('close',()=>{
        if(!closed){timer=setTimeout(connect,Math.min(30000,500*2**Math.min(attempt++,6)));timer.unref?.();}
      });
      current.addEventListener('message',message=>{
        if(closed||!this.enabled||socket!==current)return;
        let frame;try{frame=JSON.parse(message.data);}catch{return;}
        if(frame.subscription_id!==subscription||(frame.connection_id!=null&&frame.connection_id!==connectionId))return;
        attempt=0;
        if(frame.type==='snapshot'&&Number.isSafeInteger(frame.event_seq)) {
          if(state.seq!==null&&frame.event_seq<state.seq)return;
          state.seq=frame.event_seq;this.reconcileSnapshot(sid,frame.snapshot||{},true);
        } else if(frame.type==='event'||frame.type==='replay') {
          for(const envelope of frame.type==='event'?[frame.envelope]:frame.events||[]) {
            if(!Number.isSafeInteger(envelope?.seq)||envelope.seq<=(state.seq??-1)||envelope.connection_id!==connectionId)continue;
            state.seq=envelope.seq;this.applyEnvelope(sid,envelope);
          }
          if(frame.type==='replay'&&Number.isSafeInteger(frame.high_water_seq))state.seq=Math.max(state.seq??0,frame.high_water_seq);
        } else if(frame.type==='detached') {
          if(['lagged','server_shutdown'].includes(frame.reason))current.close();
          else {state.close();this.streams.delete(connectionId);}
        }
      });
    };
    connect();
  }

}

File: crates/agent-studio-core/src/adapters/codeg.rs
//! Webhook-only lifecycle. No periodic conversation/database snapshot scans.
//! Credentials at registration; keyed session metadata only after an event.
use super::*;
use crate::{merge, question, questions};
use std::time::{Duration, Instant};
pub const CODEG_EVENTS: [&str; 5] = [
    "user_prompt_sent",
    "question_request",
    "permission_request",
    "turn_complete",
    "error",
];
#[derive(Default)]
pub struct CodegHooks {
    pub url: String,
    pub registered: bool,
    pub next_attempt: Option<Instant>,
    auth: Option<(u16, String)>,
    connections: HashMap<String, String>,
    ignored_connections: std::collections::HashSet<String>,
    sequence: u64,
}
pub fn merge_codeg_webhooks(
    existing: &Value,
    owned: &[String],
    url: &str,
) -> Result<Value, String> {
    let list = existing
        .as_array()
        .ok_or("Codeg Webhook 配置无效，未覆盖")?;
    if list
        .iter()
        .any(|w| !w["url"].is_string() || !w["enabled"].is_boolean())
    {
        return Err("Codeg Webhook 配置无效，未覆盖".into());
    }
    let mut next: Vec<_> = list
        .iter()
        .filter(|w| w["url"] != url && !owned.contains(&text(&w["url"])))
        .cloned()
        .collect();
    if !url.is_empty() {
        next.push(json!({"url":url,"enabled":true}));
    }
    Ok(json!(next))
}
fn owned_urls(saved: &Value) -> Result<Vec<String>, String> {
    saved["owned"].as_array().ok_or("Codeg 注册记录无效")?
        .iter().map(|v| v.as_str().map(str::to_owned).ok_or_else(|| "Codeg 注册记录无效".to_string())).collect()
}
fn post(auth: &(u16, String), method: &str, body: Value) -> Result<Value, String> {
    ureq::AgentBuilder::new()
        .timeout(Duration::from_millis(1500))
        .build()
        .post(&format!("http://127.0.0.1:{}/api/{method}", auth.0))
        .set("Authorization", &format!("Bearer {}", auth.1))
        .send_json(body)
        .map_err(|_| "Codeg Web Service 不可用")?
        .into_json()
        .map_err(|_| "Codeg API 返回无效".to_string())
}
impl Collector {
    fn codeg_credentials(&self) -> Result<(u16, String), String> {
        let db = open(&self.paths("codeg"))?;
        let rows = query(&db,"SELECT key,value FROM app_metadata WHERE key IN ('web_service_port','web_service_token')")?;
        let value = |k: &str| {
            rows.iter()
                .find(|r| r["key"] == k)
                .map(|r| text(&r["value"]))
                .unwrap_or_default()
        };
        let port = value("web_service_port")
            .parse::<u16>()
            .ok()
            .filter(|p| *p > 0)
            .unwrap_or(3080);
        let token = value("web_service_token");
        if token.is_empty() {
            return Err("请启用 Codeg Web Service".into());
        }
        Ok((port, token))
    }
    pub fn configure_codeg_webhook(&mut self, url: String) {
        self.codeg.url = url;
        self.codeg.registered = false;
        self.codeg.next_attempt = None;
    }
    pub fn maintain_codeg_webhook(&mut self) -> Result<(), String> {
        if self.codeg.url.is_empty() {
            self.hub
                .health("codeg", "partial", "Webhook 接收入口尚未启动");
            return Ok(());
        }
        if self.codeg.registered
            || self
                .codeg
                .next_attempt
                .is_some_and(|at| Instant::now() < at)
        {
            return Ok(());
        }
        self.codeg.next_attempt = Some(Instant::now() + Duration::from_secs(60));
        let enabled = self.settings["sources"]["codeg"]["enabled"] == true && self.integration_automatic("codeg");
        let file = self.home.join(".agent-studio/codeg-webhook-native.json");
        let saved: Value = match std::fs::read(&file) {
            Ok(b) => serde_json::from_slice(&b).map_err(|_| "Codeg 注册记录无效，未覆盖")?,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => json!({"owned":[]}),
            Err(_) => return Err("Codeg 注册记录不可读".into()),
        };
        let mut owned = owned_urls(&saved)?;
        if !enabled && owned.is_empty() {
            self.codeg.registered = true;
            return Ok(());
        }
        let auth = self.codeg_credentials()?;
        let existing = post(&auth, "get_chat_event_webhooks", json!({}))?;
        let url = if enabled {
            self.codeg.url.clone()
        } else {
            String::new()
        };
        let next = merge_codeg_webhooks(&existing, &owned, &url)?;
        if enabled {
            let filter = post(&auth, "get_chat_event_filter", json!({}))?;
            let mut current: Vec<Value> = if filter.is_null() {
                CODEG_EVENTS[1..].iter().map(|e| json!(e)).collect()
            } else {
                filter
                    .as_array()
                    .filter(|a| a.iter().all(Value::is_string))
                    .cloned()
                    .ok_or("Codeg 事件配置无效")?
            };
            if CODEG_EVENTS.iter().any(|e| !current.contains(&json!(e))) {
                let channels = post(&auth, "list_chat_channels", json!({}))?;
                if channels
                    .as_array()
                    .ok_or("Codeg 推送配置无效")?
                    .iter()
                    .any(|c| c["enabled"] == true)
                    || next
                        .as_array()
                        .unwrap()
                        .iter()
                        .any(|w| w["url"] != url && w["enabled"] == true)
                {
                    return Err("Codeg 全局开关影响其他推送目标；请先在 Codeg 启用五类事件".into());
                }
                for e in CODEG_EVENTS {
                    if !current.contains(&json!(e)) {
                        current.push(json!(e));
                    }
                }
                post(&auth, "set_chat_event_filter", json!({"filter":current}))?;
            }
        }
        if !url.is_empty() && !owned.contains(&url) {
            owned.push(url.clone());
        }
        // Persist ownership before sending; uncertain network responses are safe to retry.
        atomic_json(&file, &json!({"owned":owned}))?;
        if existing != next {
            post(&auth, "set_chat_event_webhooks", json!({"webhooks":next}))?;
        }
        let verified = post(&auth, "get_chat_event_webhooks", json!({}))?;
        if verified != next { return Err("Codeg 未确认 Webhook 配置，请重试".into()); }
        atomic_json(
            &file,
            &json!({"owned":if url.is_empty(){vec![]}else{vec![url]}}),
        )?;
        self.codeg.auth = Some(auth);
        self.codeg.registered = true;
        self.hub.health(
            "codeg",
            if enabled { "ok" } else { "disabled" },
            if enabled {
                "Webhook 已注册，等待 Codeg 事件（不扫描会话）"
            } else {
                "已关闭监听"
            },
        );
        Ok(())
    }
    pub fn inspect_codeg_webhook(&self) -> Result<(&'static str, String), String> {
        let file = self.home.join(".agent-studio/codeg-webhook-native.json");
        let owned = match std::fs::read(file) {
            Ok(bytes) => serde_json::from_slice::<Value>(&bytes).map_err(|_| "Codeg 注册记录无效")?,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => json!({"owned":[]}),
            Err(e) => return Err(e.to_string()),
        };
        let owned = owned_urls(&owned)?;
        let pending = (!self.integration_automatic("codeg") || self.settings["sources"]["codeg"]["enabled"] != true) && !owned.is_empty();
        let result = self.codeg_credentials().and_then(|auth| post(&auth, "get_chat_event_webhooks", json!({})));
        match result {
            Err(e) => Ok((if pending { "pending" } else { "unavailable" }, if pending { format!("待注销；{e}，请启动 Codeg 后重试") } else {e})),
            Ok(value) => {
                let list = value.as_array().ok_or("Codeg Webhook 配置无效")?;
                let found = list.iter().any(|v| v["url"] == self.codeg.url && v["enabled"] == true);
                if pending { Ok(("pending", "接入已暂停；待重试注销".into())) }
                else if found {
                    let filter = post(&self.codeg_credentials()?, "get_chat_event_filter", json!({}))?;
                    if !CODEG_EVENTS.iter().all(|e| filter.as_array().is_some_and(|a|a.contains(&json!(e)))) { Ok(("partial", "Webhook 已注册，但事件开关不完整，请修复".into())) }
                    else { Ok(("installed", "Webhook 已注册".into())) }
                }
                else { Ok(("not_installed", "Webhook 未注册；注册需要开启监听".into())) }
            }
        }
    }
    pub fn stop_codeg_webhook(&mut self) {
        self.settings["sources"]["codeg"]["enabled"] = json!(false);
        self.codeg.registered = false;
        self.codeg.next_attempt = None;
        let _ = self.maintain_codeg_webhook();
    }
    fn codeg_metadata(&self, sid: &str) -> Result<Value, String> {
        let db = open(&self.paths("codeg"))?;
        // Parameter binding: the callback never supplies SQL identifiers or SQL text.
        let table = if db.prepare("SELECT id FROM conversation LIMIT 0").is_ok() {
            "conversation"
        } else {
            "conversations"
        };
        let mut st = db
            .prepare(&format!("SELECT * FROM {table} WHERE id=?1"))
            .map_err(|_| "Codeg 会话表不可读")?;
        let names: Vec<String> = st.column_names().iter().map(|s| s.to_string()).collect();
        let raw = st
            .query_row([sid], |r| {
                let mut out = serde_json::Map::new();
                for (i, n) in names.iter().enumerate() {
                    let v = match r.get_ref(i)? {
                        ValueRef::Text(b) => json!(String::from_utf8_lossy(b)),
                        ValueRef::Integer(n) => json!(n),
                        _ => Value::Null,
                    };
                    out.insert(n.clone(), v);
                }
                Ok(Value::Object(out))
            })
            .map_err(|_| "Codeg 会话未找到")?;
        let get = |keys: &[&str]| {
            keys.iter()
                .map(|k| raw[*k].clone())
                .find(|v| !v.is_null() && !text(v).is_empty())
                .unwrap_or(Value::Null)
        };
        let mut cwd = get(&["origin_cwd", "cwd", "workspace"]);
        if text(&cwd).is_empty() {
            if let Ok(p) = db.query_row(
                "SELECT path FROM folder WHERE id=?1",
                [text(&raw["folder_id"])],
                |r| r.get::<_, String>(0),
            ) {
                cwd = json!(p);
            }
        }
        Ok(
            json!({"cwd":cwd,"title":raw["title"],"agentType":get(&["agent_type","agent"]),"externalId":raw["external_id"],"folderId":raw["folder_id"],"isSubagent":!raw["parent_id"].is_null() || raw["kind"] == "delegate"}),
        )
    }
    pub fn ingest_codeg_hook(&mut self, p: &Value) -> bool {
        if self.settings["sources"]["codeg"]["enabled"] != true || p["source"] != "codeg" {
            return false;
        }
        let event = text(&p["event"]);
        let conn = text(&p["connection_id"]);
        if !CODEG_EVENTS.contains(&event.as_str()) || conn.trim().is_empty() || conn.len() > 256 {
            return false;
        }
        if self.codeg.ignored_connections.contains(&conn) {
            return true;
        }
        if self.codeg.auth.is_none() {
            self.codeg.auth = self.codeg_credentials().ok();
        }
        let snap = self
            .codeg
            .auth
            .as_ref()
            .and_then(|a| post(a, "acp_get_session_snapshot", json!({"connectionId":conn})).ok())
            .unwrap_or(Value::Null);
        let provisional = format!("connection:{conn}");
        let sid = if !snap["conversation_id"].is_null() {
            text(&snap["conversation_id"])
        } else {
            self.codeg
                .connections
                .get(&conn)
                .cloned()
                .unwrap_or(provisional.clone())
        };
        if sid != provisional {
            self.codeg.connections.insert(conn.clone(), sid.clone());
            self.hub.sessions.remove(&format!("codeg:{provisional}"));
        }
        if self.codeg.connections.len() > 512 {
            if let Some(k) = self.codeg.connections.keys().next().cloned() {
                self.codeg.connections.remove(&k);
            }
        }
        let meta = if sid != provisional {
            self.codeg_metadata(&sid).unwrap_or(json!({}))
        } else {
            json!({})
        };
        if meta["isSubagent"] == true {
            // Keep known children ignored during subsequent API/database outages.
            self.codeg.ignored_connections.insert(conn);
            self.hub.sessions.remove(&format!("codeg:{provisional}"));
            self.hub.sessions.remove(&format!("codeg:{sid}"));
            return true;
        }
        let ts = now();
        self.codeg.sequence += 1;
        let seq = self.codeg.sequence;
        let mut base = merge(meta, json!({"source":"codeg","sessionId":sid,"ts":ts}));
        if !snap["external_id"].is_null() {
            base["externalId"] = snap["external_id"].clone();
        }
        if !snap["folder_id"].is_null() {
            base["folderId"] = snap["folder_id"].clone();
        }
        let k = format!("codeg:{sid}");
        if event == "user_prompt_sent"
            || self
                .hub
                .sessions
                .get(&k)
                .is_none_or(|s| crate::hub::terminal(&text(&s["status"])))
        {
            let title = if text(&base["title"]).is_empty() {
                text(&p["body"])
            } else {
                text(&base["title"])
            };
            self.hub.ingest(merge(
                base.clone(),
                json!({"type":"start","roundId":format!("hook:{ts}:{seq}"),"title":title}),
            ));
        }
        if matches!(event.as_str(), "question_request" | "permission_request") {
            let ask = if !snap["pending_question"].is_null() {
                snap["pending_question"].clone()
            } else if !snap["pending_plan_approval"].is_null() {
                json!({"questions":[{"question":snap["pending_plan_approval"]["plan_markdown"],"options":[{"label":"批准"},{"label":"拒绝"}]}]})
            } else if !snap["pending_permission"].is_null() {
                let a = &snap["pending_permission"];
                json!({"questions":[{"question":a["tool_call"]["title"].as_str().unwrap_or("需要你的许可"),"options":a["options"].as_array().into_iter().flatten().map(|o|json!({"label":o["name"],"description":o["kind"]})).collect::<Vec<_>>()}]})
            } else {
                Value::Null
            };
            let pending = self.hub.sessions[&k]["pending"]
                .as_array()
                .cloned()
                .unwrap_or_default();
            for item in pending {
                self.hub.ingest(merge(
                    base.clone(),
                    json!({"type":"resolve","callId":item["id"]}),
                ));
            }
            let fields = p["fields"]
                .as_array()
                .into_iter()
                .flatten()
                .map(|f| text(&f["value"]))
                .filter(|s| !s.is_empty())
                .collect::<Vec<_>>()
                .join("\n");
            let message = if !ask.is_null() {
                question(&ask)
            } else if !fields.is_empty() {
                fields
            } else {
                text(&p["body"])
            };
            let call = ask["question_id"]
                .as_str()
                .or(ask["request_id"].as_str())
                .map(str::to_owned)
                .unwrap_or(format!("codeg:{conn}:{seq}"));
            self.hub.ingest(merge(base,json!({"type":"wait","callId":call,"tool":if event=="question_request"{"ask"}else{"permission"},"text":message,"questions":questions(&ask)})));
        } else if matches!(event.as_str(), "turn_complete" | "error") {
            self.hub.ingest(merge(
                base,
                json!({"type":"end","status":if event=="error"{"error"}else{"done"}}),
            ));
        }
        self.hub.health(
            "codeg",
            "ok",
            "已收到 Codeg Webhook；回答后的状态等待下一条事件更新",
        );
        true
    }
}
